### ISSUE TITLE
FIRE Relative Positional Encodings

### ISSUE THREAD

**Issue by kaddu341** [pytorch/torchtune#2325]
Hi,

I'm currently working on the length generalization capabilities of transformers. As shown by Zhou et al. (https://arxiv.org/abs/2402.09371), FIRE positional encodings are excellent for this purpose as they can yield generalization results up to 2.5x the input length (in combination with other techniques).

FIRE, which stands for Functional Interpolation for Relative Positional Encodings, was introduced by Li et al. (https://arxiv.org/pdf/2310.04418).

I am planning to implement the algorithm from this paper myself, but I thought it would be useful if I could turn it into a PyTorch module so that others can benefit too. (I [posted this](https://github.com/pytorch/pytorch/issues/145851) originally in the Pytorch Core repo, but they suggested to bring it here). Therefore, I am proposing to add this feature to the torchtune library.

Please let me know what you think!

There are many other positional encoding types (sinusoidal, RoPE, learned, etc.), but for the specific task of length generalization, FIRE seems to be the most suitable based on several papers, which is why I am proposing this feature addition.

Like other relative attention mechanisms, FIRE introduces positional information in the attention layers rather than adding it to the input.

Here is a screenshot of some evaluation results for FIRE from the original paper (Li et al., 2024):

<img width="890" alt="Image" src="https://github.com/user-attachments/assets/614c8780-4559-47a3-be09-02ce093718e7" />

**Comment by kaddu341** [comment]
Sure!

### Context:
The length generalization of transformers is a problem currently generating much interest in the machine learning community. Generalization performance depends largely on the type of positional encoding used. Over the years, several new positional encoding schemes have been proposed, including FIRE (Li et al., 2023/2024) which has been shown to produce impressive generalization results (Zhou et al., 2024).

### Motivation:
For many tasks, the length of the training sequences that we have available is less than the length of the sequences we would like to run inference on. This is especially important for physical simulations, but also for certain natural language processing (NLP) tasks.

### Details:
Like most other relative positional encoding schemes, FIRE adds a bias matrix to the standard self-attention computation (before applying softmax) as follows:

<img width="403" alt="Image" src="https://github.com/user-attachments/assets/2cb9e7c9-c7e2-470e-acbf-b5711bb5244c" />

where $B$ is calculated by:

<img width="958" alt="Image" src="https://github.com/user-attachments/assets/4d041d1d-97dc-49f5-9ae5-8e359e75ec9a" />

In code, this can be expressed as follows (please note this is not production code and may have typos; it is just to give an idea of the algorithm)
```
# class to implement a single self-attention head
class FIRE_AttentionHead(nn.Module):
    def __init__(self, dim_model, kdim, hidden_size):
        super().__init__()
        self.kdim = kdim

        # initialize parameter matrices
        self.W_q = nn.Linear(dim_model, kdim, bias=False)
        self.W_k = nn.Linear(dim_model, kdim, bias=False)
        self.W_v = nn.Linear(dim_model, kdim, bias=False)

        # initialize learnable scalars
        self.c = nn.Parameter(torch.tensor(1.0))
        self.L = nn.Parameter(torch.tensor(2.0))

        # initialize learnable continuous function
        self.f_theta = nn.Sequential(
            nn.Linear(1, hidden_size),
            nn.GELU(),
            nn.Linear(hidden_size, hidden_size),
            nn.GELU(),
            nn.Linear(hidden_size, 1)
        )
    
    # concave function to amplify differences among local positions
    def phi(self, c, x):
        return torch.log(c*x + 1)
    
    def forward(self, src):
        # Assuming src has shape (batch_size, seq_length, dim_model)
        batch_size = src.shape[0]
        seq_length = src.shape[1]

        # constrain c to be > 0
        c = torch.nn.functional.softplus(self.c)

        # compute bias matrix
        B = torch.zeros(seq_length, seq_length)
        for i in range(1, seq_length):
            for j in range(0, i):
                B[i, j] = self.phi(c, i - j) / self.phi(c, max(self.L, i + 1))
        # make it antisymmetric:
        B = B - B.T
        # apply MLP to bias matrix
        B = self.f_theta(B.unsqueeze(2)).squeeze(2)
        # repeat B for batch_size
        B = B.repeat(batch_size, 1, 1)

        # get Query, Key, and Value matrices for each sequence
        Q = self.W_q(src)
        K = self.W_k(src)
        V = self.W_v(src)

        # calculate attention scores
        K_T = torch.transpose(K, 1, 2)
        attn_logits  = torch.bmm(Q, K_T) + B    # / np.sqrt(self.kdim)
        attn_weights = torch.nn.functional.softmax(attn_logits, dim=-1)
        attn_outputs = torch.bmm(attn_weights, V)
        return attn_outputs

# class to implement multi-head self-attention with relative positional encodings according to the FIRE paper
class FIRE_SelfAttention(nn.Module):
    def __init__(self, dim_model, num_heads, hidden_size):
        super().__init__()

        # make sure num_heads divides dim_model:
        assert dim_model % num_heads == 0, "Number of heads must divide dimension of model"

        # compute kdim = vdim
        kdim = dim_model // num_heads

        # initialize attention heads
        self.attention_heads = nn.ModuleList([FIRE_AttentionHead(dim_model, kdim, hidden_size) for _ in range(num_heads)])

        # final linear layer
        self.W_o = nn.Linear(dim_model, dim_model, bias=False)

    def forward(self, src):
        # src should have shape (batch_size, seq_length, dim_model)
        # Pass src through the attention heads
        attn_results = [attn_head(src) for attn_head in self.attention_heads]
        # concatenate results
        attn_results = torch.cat(attn_results, dim=-1)
        # pass through final linear layer
        return self.W_o(attn_results)
```

I am planning to make this class inherit from nn.Module and basically function as a drop-in replacement for nn.MultiHeadAttention.

**Comment by acisseJZhong** [comment]
Hi @kaddu341 thanks so much for drafting the RFC. It's very clear. Would you mind post it as a PR so that other people can comment under?

**Comment by kaddu341** [comment]
Sure, I would love to! Let me just clean up the code a bit though.

Also, it's my first time making a pull request - do I just add the code to the relevant file and open the PR or is there something else (testing etc.) I need to do as well?

Thanks for the help!

**Comment by pbontrager** [comment]
Can you give more context on why this needs to be implemented as a new attention layer instead of just a positional encoding? The attention layers have a lot more considerations as they need to work with kvcache during inference and since it's used everywhere the MHA module is already fairly optimized.

Does FIRE need to be fine-tuned on or is it dropped in at inference? Does it work if you use it on a pre-trained model or does the model have to be pre-trained with it? If I fine-tuned a model with FIRE with the custom attention layer, I don't think you'd be able to run the model in inference mode on vllm as it has it's own paged attention layer for fast inference.

Sorry for this barrage of questions but I need to understand how this will be used within torchtune.

**Comment by pbontrager** [comment]
I understand all the points you're giving, and really appreciate the detailed writeup and planning work you did so far. I'm sorry this got so far before a proper discussion about the purpose for this module, but the scope of torchtune is as a post-training library. All the models we have implemented are existing pre-trained models which we then provide recipes to fine-tune on. There is [torchtitan](https://github.com/pytorch/torchtitan) for pre-training. I'm sorry again that this got so far, but I believe this module is out of scope for torchtune.

**Comment by kaddu341** [comment]
No problem, I understand.

Would you suggest opening an issue/pull request for this module on torchtitan, or is there another library that is better suited for this purpose?
